### PR TITLE
[dynamo][higher order op] Support nn.Module calls

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102203
* #102123
* __->__ #102022

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -511,10 +515,14 @@ def register_attr_or_module(
         assert not isinstance(source, ParamBufferSource)
 
         if isinstance(target, torch.Tensor):
-            if len(self.tracers) > 1:
-                unimplemented(
-                    "accessing attribute of nn.Module inside HigherOrderOperator"
-                )
+            tracer = self.current_tracer
+            if not self.is_root_tracer():
```
Comment (tugsbayasgalan) [inline torch/_dynamo/output_graph.py:519]: Can this handle nested HigherOrderOperator calls?

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -511,10 +515,14 @@ def register_attr_or_module(
...
+                # For higher order ops, we don't want to insert the get_attr in
+                # innermost graph. Instead, we want to raise the params/buffers
+                # as inputs to the higher-order graph, and register them as
+                # get_attrs in the root tracer.
+                tracer = self.root_tracer
```
Comment (voznesenskym) [inline torch/_dynamo/output_graph.py:524]: Hmm I am not sure about this - we already do lifting through `lift_tracked_freevar_to_input` - why doesn't this go through that?

File: torch/_dynamo/output_graph.py [hunk 3]
```diff
@@ -1003,11 +1011,12 @@ def __init__(self, output_graph, parent=None):
...
-        # of the graph. If we are tracing a HigherOrderOperator's body_fn,
-        # then we need to keep track of this so we can rewrite the
-        # HigherOrderOperator call using the traced body_fn.
-        self.lifted_freevars = set({})
+        # A list of previously free variables that we lifted to being inputs of
+        # the graph. If we are tracing a HigherOrderOperator's body_fn, then we
+        # need to keep track of this so we can rewrite the HigherOrderOperator
+        # call using the traced body_fn. This is a list (instead of set) so that
+        # we can maintain the order of args for the HigherOrderOperator call.
+        self.lifted_freevars = list()
```
Comment (voznesenskym) [inline torch/_dynamo/output_graph.py:1019]: Use dict and have keys w/ None values, that gives you both ordering and uniqueness.

File: torch/_dynamo/output_graph.py (hunk 1 above)
Comment (anijain2305) [reply torch/_dynamo/output_graph.py:519]: Good point. Maybe not.

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (anijain2305) [reply torch/_dynamo/output_graph.py:524]: The main difference is that we don't want to raise the parameters as inputs of the outer graph. This way, we don't change the calling convention for downstream users.

Comment (anijain2305) [reply torch/_dynamo/output_graph.py:524]: Note that Dynamo will still call `lift_tracked_freevar_to_input` for the inner graph. The only difference is what happens at the root.  If not a parameter, this is what happens today for higher order ops

https://github.com/pytorch/pytorch/blob/2fa1b563da4b14c29e1a8572cefc79280caf7764/torch/_dynamo/variables/builder.py#L814-L824


This is resulting graph

~~~
 ===== __compiled_fn_0 =====
 <eval_with_key>.1 class GraphModule(torch.nn.Module):
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_

        # File: /scratch/anijain/work/pytorch/torch/nn/modules/linear.py:114, code: return F.linear(input, self.weight, self.bias)
        g__mod___linear_weight = self.G__mod___linear_weight
        g__mod___linear_bias = self.G__mod___linear_bias

        # File: /scratch/anijain/work/pytorch/examples/higher.py:39, code: return torch.utils.checkpoint.checkpoint(mod, x)
        wrap_body_0 = self.wrap_body_0
        wrap_activation_checkpoint = torch._higher_order_ops.wrap.wrap_activation_checkpoint(wrap_body_0, l_x_, g__mod___linear_weight, g__mod___linear_bias);  wrap_body_0 = l_x_ = g__mod___linear_weight = g__mod___linear_bias = None
        return (wrap_activation_checkpoint,)

    class GraphModule(torch.nn.Module):
        def forward(self, l_x_, g__mod___linear_weight, g__mod___linear_bias):
            # File: /scratch/anijain/work/pytorch/torch/nn/modules/linear.py:114, code: return F.linear(input, self.weight, self.bias)
            linear = torch._C._nn.linear(l_x_, g__mod___linear_weight, g__mod___linear_bias);  l_x_ = g__mod___linear_weight = g__mod___linear_bias = None

            # File: /scratch/anijain/work/pytorch/examples/higher.py:34, code: return torch.sigmoid(self.linear(x))
            sigmoid = torch.sigmoid(linear);  linear = None
            return sigmoid
~~~

File: torch/_dynamo/output_graph.py (hunk 1 above)
Comment (anijain2305) [reply torch/_dynamo/output_graph.py:519]: Can you give me an example on how to test this? I think it should work. We will still raise the params/buffers as inputs for all the inner graphs, and the outermost_graph will make the right connection.

File: torch/_dynamo/variables/functions.py [hunk 4]
```diff
@@ -308,7 +308,13 @@ def call_function(
         # since we ensure `forward` of allowed modules can be traced by AOT safely.
         # Note this is not only for allowed modules, as user customized modules can extend from
         # allowed modules but using parent's `forward` method, which is also covered by this branch.
-        if isinstance(self.obj, variables.NNModuleVariable):
+
+        # If we are tracing the higher order op, we want Dynamo to step inside
+        # the module call so that Dynamo can see the underlying parameters and
+        # buffers and raise them as inputs to the graph.
+        if tx.output.is_root_tracer() and isinstance(
```
Comment (voznesenskym) [inline torch/_dynamo/variables/functions.py:315]: This effectively delegates it to the super() for a non root tracer, right?

File: torch/_dynamo/output_graph.py (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/output_graph.py:519]: Yeah i was just thinking about nested control flow

File: torch/_dynamo/variables/functions.py (hunk 4 above)
Comment (anijain2305) [reply torch/_dynamo/variables/functions.py:315]: Correct, which in turn allows Dynamo to trace the innards of `forward` method.

**Review by zou3519 — COMMENTED** [review]
This looks great! Code LGTM, I had some minor comments on additional things to test

File: torch/_dynamo/variables/functions.py (hunk 4 above)
Comment (anijain2305) [reply torch/_dynamo/variables/functions.py:315]: Fair enough. Let me add comments to say how `super` means inlining.

File: torch/_dynamo/output_graph.py (hunk 3 above)
Comment (anijain2305) [reply torch/_dynamo/output_graph.py:1026]: Done

File: test/dynamo/test_higher_order_ops.py [hunk 5]
```diff
@@ -544,6 +567,60 @@ def f(x):
...
+
+        model = ToyModel()
+        forward_handles = {}
+        activations = dict()
+
+        def save_activations(mod, inp, out):
+            activations[name] = inp
+
+        for name, module in model.named_children():
+            forward_handles[name] = module.register_forward_hook(save_activations)
```
Comment (zou3519) [inline test/dynamo/test_higher_order_ops.py:607]: Can we add a test for `module.register_full_backward_hook`? I would expect this to cause the entire wrap() to fall back to eager mode (because we traced out the module, there is no way for the full backward hook to actually fire).

File: test/dynamo/test_higher_order_ops.py [hunk 6]
```diff
@@ -544,6 +567,60 @@ def f(x):
...
+
+        @torch.compile(backend="eager")
+        def fn(x):
+            return wrap(lambda x: model(x), x)
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply test/dynamo/test_higher_order_ops.py:611]: Good catch! I did eyeball it last night, we were not falling back. Let me add `fullgraph`.

**Review by zou3519 — APPROVED** [review]
LGTM, see comments
